Mathlib/CategoryTheory/FiberedCategory/HasFibers.lean

```diff
@@ -0,0 +1,248 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Calle Sönne, Paul Lezeau
+-/
+
+import Mathlib.CategoryTheory.FiberedCategory.Fiber
+import Mathlib.CategoryTheory.FiberedCategory.Fibered
+
+/-!
+
+# Fibers of functors
+
+In this file we introduce a typeclass `HasFibers` for a functor `p : 𝒳 ⥤ 𝒮`, consisting of:
+- A collection of categories `Fib S` for every `S` in `𝒮` (the fiber categories)
+- Functors `ι : Fib S ⥤ 𝒳` such that `ι ⋙ p = const (Fib S) S
+- The induced functor `Fib S ⥤ Fiber p S` is an equivalence.
+
+We also provide a default `HasFibers` instance, which uses the standard fibers `Fiber p S`
+(see Fiber.lean). This makes it so that any result proven about `HasFibers` can be used for the
+standard fibers as well.
+
+The reason for introducing this typeclass is that in practice, when working with (pre)fibered
+categories one often already has a collection of categories `Fib S` for every `S` that are
+equivalent to the fibers `Fiber p S`. One would then like to use these categories `Fib S` directly,
+instead of working through this equivalence of categories. By developing an API for the `HasFibers`
+typeclass, this will be possible.
+
+Here is an example of when this typeclass is useful. Suppose we have a presheaf of types
+`F : 𝒮ᵒᵖ ⥤ Type _`. The associated fibered category then has objects `(S, a)` where `S : 𝒮` and `a`
+is an element of `F(S)`. The fiber category `Fiber p S` is then equivalent to the discrete category
+`Fib S` with objects `a` in `F(S)`. In this case, the `HasFibers` instance is given by the
+categories `F(S)` and the functor `ι` sends `a : F(S)` to `(S, a)` in the fibered category. See
+`Presheaf.lean` for more details.
```

Comment:
It seems this example is not yet in mathlib:
```suggestion
categories `F(S)` and the functor `ι` sends `a : F(S)` to `(S, a)` in the fibered category.
```

Reply:
Thanks for catching that! Yeah I only have that code on the stacks project.